components/common/hooks/usePreference.ts

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {useDispatch, useSelector} from 'react-redux';
+
+import {PreferenceType} from '@mattermost/types/preferences';
+
+import {savePreferences} from 'mattermost-redux/actions/preferences';
+import {getCurrentUserId} from 'mattermost-redux/selectors/entities/common';
+import {getMyPreferences} from 'mattermost-redux/selectors/entities/preferences';
+import {getPreferenceKey} from 'mattermost-redux/utils/preference_utils';
+
+export default function usePreference(category: string, name: string): [string | undefined, (value: string) => void] {
```

Comment:
@neallred I noticed you recently added `hooks/useSavePreference.ts`. Should this be defined in the same file? I also noticed you're using `useCallback`, whereas I did not here

Reply:
Yeah, I've been adding a bunch of hooks lately, and I couldn't remember whose awesome preference hook I had seen in a PR. So I ended up more or less duplicating what you have here. There's some nice things in both of them.

I think it makes sense combining them into the same file. It's not for a single hook anymore, so maybe a good filename would be `components/common/hooks/usePreference.ts`

Yours is so general and useful that I can see it being used a LOT, so it might be worth adding a `useCallback` and a `useMemo` just so that in components that are in the hot path for rendering, we don't inadvertently cause re-render issues. Something like

```tsx
+++ b/components/common/hooks/usePreference.ts
@@ -1,6 +1,8 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.

+import {useMemo, useCallback} from 'react';
+
 import {useDispatch, useSelector} from 'react-redux';

 import {PreferenceType} from '@mattermost/types/preferences';
@@ -19,7 +21,7 @@ export default function usePreference(category: string, name: string): [string |
     const key = getPreferenceKey(category, name);
     const preference = preferences[key];

-    const setPreference = (value: string) => {
+    const setPreference = useCallback((value: string) => {
         const preference: PreferenceType = {
             category,
             name,
@@ -27,7 +29,7 @@ export default function usePreference(category: string, name: string): [string |
             value,
         };
         dispatch(savePreferences(userId, [preference]));
-    };
+    }, [userId]);

-    return [preference?.value, setPreference];
+    return useMemo(() => ([preference?.value, setPreference]), [preference?.value, setPreference]);
 }
```